File: src/merge.rs
use crate::cli::MergeArgs;
use crate::constants::{DATE_FORMAT_COMPACT, DATE_FORMAT_DASHED, MAX_FILE_SIZE};
use chrono::{NaiveDate, NaiveTime};
use glob::{glob, GlobError, PatternError};
use std::fs;
use std::io::Write;
use std::path::{Path, PathBuf};
use tempfile::NamedTempFile;

#[derive(Debug, Clone)]
pub struct MergeRequest {
    pub patterns: Vec<String>,
    pub output: Option<PathBuf>,
    pub no_delete: bool,
}

impl From<MergeArgs> for MergeRequest {
    fn from(args: MergeArgs) -> Self {
        MergeRequest {
            patterns: args.patterns,
            output: args.output,
            no_delete: args.no_delete,
        }
    }
}

#[derive(Debug)]
pub struct MergeOutcome {
    pub files: Vec<PathBuf>,
    pub output_path: PathBuf,
}

#[derive(Debug, thiserror::Error)]
pub enum MergeError {
    #[error("Invalid glob pattern '{pattern}': {source}")]
    InvalidGlobPattern {
        pattern: String,
        #[source]
        source: PatternError,
    },
    #[error("No files matched pattern '{0}'")]
    NoMatches(String),
    #[error("Failed to read glob matches: {0}")]
    GlobIteration(#[from] GlobError),
    #[error("Unrecognized transcript filename '{0}'")]
    UnrecognizedFilename(String),
    #[error("Files correspond to multiple dates; supply --output to choose the destination")]
    MixedDates,
    #[error("Unable to determine an output filename; rerun with --output <file>")]
    UndeterminedDate,
    #[error("I/O error: {0}")]
    Io(#[from] std::io::Error),
    #[error("File too large: {path} ({size} bytes exceeds maximum of {max} bytes)")]
    FileTooLarge { path: String, size: u64, max: u64 },
}

/// Execute the merge operation on transcript files.
pub fn execute(request: &MergeRequest) -> Result<MergeOutcome, MergeError> {
    let mut collected = Vec::new();

    for pattern in &request.patterns {
        let mut matches_found = false;
        let entries = glob(pattern).map_err(|err| MergeError::InvalidGlobPattern {
            pattern: pattern.clone(),
            source: err,
        })?;
        for entry in entries {
            let path = entry?;
            matches_found = true;
            collected.push(path);
        }

        if !matches_found {
            return Err(MergeError::NoMatches(pattern.clone()));
        }
    }

    // Check file sizes before processing to prevent OOM
    for path in &collected {
        let metadata = fs::metadata(path)?;
        if metadata.len() > MAX_FILE_SIZE {
            return Err(MergeError::FileTooLarge {
                path: path.display().to_string(),
                size: metadata.len(),
                max: MAX_FILE_SIZE,
            });
        }
    }

    let mut descriptors: Vec<(PathBuf, FileSortKey)> = collected
        .into_iter()
        .map(|path| {
            let key = FileSortKey::from_path(&path)?;
            Ok((path, key))
        })
        .collect::<Result<_, MergeError>>()?;

    descriptors.sort_by(|a, b| a.1.cmp(&b.1));

    let mut ordered = Vec::new();
    for (path, _) in &descriptors {
        if ordered
            .last()
            .map(|existing: &PathBuf| existing == path)
            .unwrap_or(false)
        {
            continue;
        }
        ordered.push(path.clone());
    }

    let output_path = determine_output_path(&ordered, &descriptors, request)?;

    // Canonicalize output path for reliable comparison
    // If output doesn't exist yet, we'll compare against the intended path
    let output_canonical = output_path.canonicalize().unwrap_or_else(|_| output_path.clone());

    // Filter out the output path from sources to prevent self-deletion
    let sources_to_merge: Vec<PathBuf> = ordered
        .iter()
        .filter(|p| {
            let p_canonical = p.canonicalize().unwrap_or_else(|_| (*p).clone());
            p_canonical != output_canonical
        })
        .cloned()
        .collect();

    write_merged_file(&sources_to_merge, &output_path)?;
    if !request.no_delete {
        delete_sources(&sources_to_merge, &output_path)?;
    }

    Ok(MergeOutcome {
        files: sources_to_merge,
        output_path,
    })
}

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
struct FileSortKey {
    date: Option<NaiveDate>,
    start: NaiveTime,
}

impl FileSortKey {
    fn from_path(path: &Path) -> Result<Self, MergeError> {
        let filename = path
            .file_stem()
            .and_then(|s| s.to_str())
            .ok_or_else(|| MergeError::UnrecognizedFilename(path.display().to_string()))?;

        // Try flat format first: YYYYMMDD_HHMMSS_HHMMSS (more specific pattern)
        if Self::looks_like_flat_format(filename) {
            return Self::parse_flat(path, filename);
        }

        // Try nested format: HHMMSS-HHMMSS (6 digits, dash, 6 digits)
        if Self::looks_like_nested_format(filename) {
            return Self::parse_nested(path, filename);
        }

        Err(MergeError::UnrecognizedFilename(path.display().to_string()))
    }

    /// Check if filename matches flat format: YYYYMMDD_HHMMSS_HHMMSS
    fn looks_like_flat_format(filename: &str) -> bool {
        let parts: Vec<&str> = filename.split('_').collect();
        if parts.len() != 3 {
            return false;
        }
        // Date part: 8 digits
        if parts[0].len() != 8 || !parts[0].chars().all(|c| c.is_ascii_digit()) {
            return false;
        }
        // Start time: 6 digits
        if parts[1].len() != 6 || !parts[1].chars().all(|c| c.is_ascii_digit()) {
            return false;
        }
        // End time: 6 digits
        if parts[2].len() != 6 || !parts[2].chars().all(|c| c.is_ascii_digit()) {
            return false;
        }
        true
    }

    /// Check if filename matches nested format: HHMMSS-HHMMSS
    fn looks_like_nested_format(filename: &str) -> bool {
        let parts: Vec<&str> = filename.split('-').collect();
        if parts.len() != 2 {
            return false;
        }
        // Start time: 6 digits
        if parts[0].len() != 6 || !parts[0].chars().all(|c| c.is_ascii_digit()) {
            return false;
        }
        // End time: 6 digits
        if parts[1].len() != 6 || !parts[1].chars().all(|c| c.is_ascii_digit()) {
            return false;
        }
        true
    }

    fn parse_nested(path: &Path, filename: &str) -> Result<Self, MergeError> {
        let mut segments = filename.split('-');
        let start_segment = segments
            .next()
            .ok_or_else(|| MergeError::UnrecognizedFilename(path.display().to_string()))?;

        let start = parse_time_digits(start_segment)
            .ok_or_else(|| MergeError::UnrecognizedFilename(path.display().to_string()))?;

        let date = parse_date_from_path(path);

        Ok(FileSortKey { date, start })
    }

    fn parse_flat(path: &Path, filename: &str) -> Result<Self, MergeError> {
        let mut parts = filename.split('_');
        let date_part = parts
            .next()
            .ok_or_else(|| MergeError::UnrecognizedFilename(path.display().to_string()))?;
        let start_part = parts
            .next()
            .ok_or_else(|| MergeError::UnrecognizedFilename(path.display().to_string()))?;

        let date = NaiveDate::parse_from_str(date_part, DATE_FORMAT_COMPACT)
            .map_err(|_| MergeError::UnrecognizedFilename(path.display().to_string()))?;
        let start = parse_time_digits(start_part)
            .ok_or_else(|| MergeError::UnrecognizedFilename(path.display().to_string()))?;

        Ok(FileSortKey {
            date: Some(date),
            start,
        })
    }
}

fn parse_time_digits(value: &str) -> Option<NaiveTime> {
    if value.len() != 6 {
        return None;
    }
    let hour = value.get(0..2)?.parse::<u32>().ok()?;
    let minute = value.get(2..4)?.parse::<u32>().ok()?;
    let second = value.get(4..6)?.parse::<u32>().ok()?;
    NaiveTime::from_hms_opt(hour, minute, second)
}

fn parse_date_from_path(path: &Path) -> Option<NaiveDate> {
    extract_nested_day_directory(path).map(|(_, date)| date)
}

fn extract_nested_day_directory(path: &Path) -> Option<(PathBuf, NaiveDate)> {
    let day_dir = path.parent()?;
    let day_name = day_dir.file_name()?.to_str()?;

    if let Ok(date) = NaiveDate::parse_from_str(day_name, DATE_FORMAT_DASHED) {
        return Some((day_dir.to_path_buf(), date));
    }

    let month_dir = day_dir.parent()?;
    let year_dir = month_dir.parent()?;
    let month_name = month_dir.file_name()?.to_str()?;
    let year_name = year_dir.file_name()?.to_str()?;

    if year_name.len() == 4
        && month_name.len() == 2
        && day_name.len() == 2
        && year_name
            .chars()
            .chain(month_name.chars())
            .chain(day_name.chars())
            .all(|c| c.is_ascii_digit())
    {
        if let Ok(date) =
            NaiveDate::parse_from_str(&format!("{year_name}-{month_name}-{day_name}"), DATE_FORMAT_DASHED)
        {
            return Some((day_dir.to_path_buf(), date));
        }
    }

    None
}

fn detect_common_nested_directory(paths: &[PathBuf]) -> Option<(PathBuf, NaiveDate)> {
    let mut candidate: Option<(PathBuf, NaiveDate)> = None;
    for path in paths {
        let info = extract_nested_day_directory(path)?;
        match &candidate {
            Some((dir, date)) if dir == &info.0 && *date == info.1 => continue,
            Some(_) => return None,
            None => candidate = Some(info),
        }
    }
    candidate
}

fn determine_output_path(
    ordered: &[PathBuf],
    descriptors: &[(PathBuf, FileSortKey)],
    request: &MergeRequest,
) -> Result<PathBuf, MergeError> {
    if let Some(custom) = &request.output {
        return Ok(custom.clone());
    }

    if let Some((day_dir, date)) = detect_common_nested_directory(ordered) {
        return Ok(day_dir.join(format!("{}.txt", date.format(DATE_FORMAT_DASHED))));
    }

    let mut selected_date: Option<NaiveDate> = None;
    for (_, key) in descriptors {
        if let Some(date) = key.date {
            if let Some(existing) = selected_date {
                if existing != date {
                    return Err(MergeError::MixedDates);
                }
            } else {
                selected_date = Some(date);
            }
        }
    }

    if let Some(date) = selected_date {
        let base_dir = ordered
            .first()
            .and_then(|path| path.parent())
            .map(|p| p.to_path_buf())
            .unwrap_or_else(|| PathBuf::from("."));
        return Ok(base_dir.join(format!("{}.txt", date.format(DATE_FORMAT_DASHED))));
    }

    Err(MergeError::UndeterminedDate)
}

fn write_merged_file(files: &[PathBuf], output_path: &Path) -> Result<(), MergeError> {
    let mut merged = String::new();
    for (idx, path) in files.iter().enumerate() {
        let segment = fs::read_to_string(path)?;
        merged.push_str(&segment);
        if idx + 1 != files.len() && !merged.ends_with('\n') {
            merged.push('\n');
        }
    }

    if let Some(parent) = output_path.parent() {
        if !parent.as_os_str().is_empty() {
            fs::create_dir_all(parent)?;
        }
    }

    // Atomic write: write to temp file then rename
    atomic_write(output_path, merged.as_bytes())?;
    Ok(())
}

/// Write content atomically by writing to a temp file and renaming.
/// This prevents partial writes on crash.
fn atomic_write(path: &Path, content: &[u8]) -> std::io::Result<()> {
    let parent = path.parent().unwrap_or(Path::new("."));
    let mut temp = NamedTempFile::new_in(parent)?;
    temp.write_all(content)?;
    temp.flush()?;
    temp.persist(path).map_err(|e| e.error)?;
    Ok(())
}

fn delete_sources(files: &[PathBuf], output_path: &Path) -> Result<(), MergeError> {
    let output_canonical = output_path.canonicalize().unwrap_or_else(|_| output_path.to_path_buf());
    for path in files {
        // Double-check: never delete the output file
        let path_canonical = path.canonicalize().unwrap_or_else(|_| path.clone());
        if path_canonical == output_canonical {
            continue;
        }
        fs::remove_file(path)?;
    }
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;
    use assert_fs::prelude::*;
    use std::fs;
    use std::path::Path;

    #[test]
    fn sorts_nested_files() {
        let temp = assert_fs::TempDir::new().unwrap();
        let day_dir = temp.child("2025/01/27");
        day_dir.create_dir_all().unwrap();
        day_dir
            .child("112256-162256.txt")
            .write_str("late segment\n")
            .unwrap();
        day_dir
            .child("061901-111901.txt")
            .write_str("early segment\n")
            .unwrap();

        let request = MergeRequest {
            patterns: vec![day_dir.path().join("*.txt").to_string_lossy().into()],
            output: None,
            no_delete: false,
        };
        let outcome = execute(&request).unwrap();
        assert_eq!(outcome.files.len(), 2);
        assert!(
            outcome.files[0].ends_with("061901-111901.txt"),
            "expected earliest file first"
        );
        assert!(
            outcome.files[1].ends_with("112256-162256.txt"),
            "expected later file second"
        );
        assert!(
            outcome
                .output_path
                .ends_with(Path::new("2025/01/27/2025-01-27.txt")),
            "unexpected output path {:?}",
            outcome.output_path
        );
        let merged = fs::read_to_string(outcome.output_path).unwrap();
        assert_eq!(merged, "early segment\nlate segment\n");
    }

    #[test]
    fn sorts_flat_files() {
        let temp = assert_fs::TempDir::new().unwrap();
        temp.child("20250127_112256_162256.txt")
            .write_str("late\n")
            .unwrap();
        temp.child("20250127_061901_111901.txt")
            .write_str("early\n")
            .unwrap();

        let request = MergeRequest {
            patterns: vec![temp.path().join("20250127_*.txt").to_string_lossy().into()],
            output: None,
            no_delete: false,
        };
        let outcome = execute(&request).unwrap();
        assert_eq!(outcome.files.len(), 2);
        assert!(
            outcome.files[0].ends_with("20250127_061901_111901.txt"),
            "expected earliest file first"
        );
        assert!(
            outcome.files[1].ends_with("20250127_112256_162256.txt"),
            "expected later file second"
        );
        assert!(
            outcome.output_path.ends_with(Path::new("2025-01-27.txt")),
            "unexpected output path {:?}",
            outcome.output_path
        );
        let merged = fs::read_to_string(outcome.output_path).unwrap();
        assert_eq!(merged, "early\nlate\n");
    }

    #[test]
    fn rejects_non_transcript_filenames() {
        // Test that random files with dashes or underscores are properly rejected
        assert!(!FileSortKey::looks_like_nested_format("meeting-notes"));
        assert!(!FileSortKey::looks_like_nested_format("2025-01-27")); // date format
        assert!(!FileSortKey::looks_like_nested_format("abc123-def456"));
        assert!(!FileSortKey::looks_like_flat_format("notes_about_meeting"));
        assert!(!FileSortKey::looks_like_flat_format("2025_01_27"));

        // Valid formats should be accepted
        assert!(FileSortKey::looks_like_nested_format("112256-162256"));
        assert!(FileSortKey::looks_like_flat_format("20250127_112256_162256"));
    }

    #[test]
    fn excludes_output_file_from_merge_sources() {
        // Test that the output file won't be deleted even if explicitly listed as a source
        let temp = assert_fs::TempDir::new().unwrap();
        let day_dir = temp.child("2025/01/27");
        day_dir.create_dir_all().unwrap();
        day_dir
            .child("112256-162256.txt")
            .write_str("segment 1\n")
            .unwrap();
        day_dir
            .child("061901-111901.txt")
            .write_str("segment 2\n")
            .unwrap();

        // Use explicit file paths matching the HHMMSS-HHMMSS pattern
        let file1 = day_dir.path().join("061901-111901.txt");
        let file2 = day_dir.path().join("112256-162256.txt");
        let output_file = day_dir.path().join("merged.txt");

        let request = MergeRequest {
            patterns: vec![
                file1.to_string_lossy().into(),
                file2.to_string_lossy().into(),
            ],
            output: Some(output_file.clone()),
            no_delete: false,
        };
        let outcome = execute(&request).unwrap();

        // Source files should be deleted
        assert!(!file1.exists());
        assert!(!file2.exists());

        // Output file should exist
        assert!(outcome.output_path.exists());
        let merged = fs::read_to_string(&outcome.output_path).unwrap();
        assert_eq!(merged, "segment 2\nsegment 1\n");
    }

    #[test]
    fn rejects_non_matching_files_in_glob() {
        // Test that files not matching transcript patterns cause errors
        let temp = assert_fs::TempDir::new().unwrap();
        let day_dir = temp.child("2025/01/27");
        day_dir.create_dir_all().unwrap();
        day_dir
            .child("112256-162256.txt")
            .write_str("segment\n")
            .unwrap();
        day_dir
            .child("notes.txt")
            .write_str("random notes\n")
            .unwrap();

        // Using *.txt should fail because notes.txt doesn't match
        let request = MergeRequest {
            patterns: vec![day_dir.path().join("*.txt").to_string_lossy().into()],
            output: None,
            no_delete: false,
        };
        let result = execute(&request);
        assert!(matches!(result, Err(MergeError::UnrecognizedFilename(_))));
    }
}

File: src/merge.ts
/**
 * Merge command implementation - merges multiple transcript segments.
 */

import * as fs from 'node:fs';
import * as path from 'node:path';
import { glob } from 'glob';
import { MAX_FILE_SIZE, DATE_COMPACT_REGEX } from './constants.js';
import { atomicWrite } from './update.js';
import { type DateOnly, formatDateDashed } from './transcript.js';

/** Request parameters for the merge command. */
export interface MergeRequest {
  patterns: string[];
  output?: string;
  noDelete: boolean;
}

/** Result of the merge operation. */
export interface MergeOutcome {
  files: string[];
  outputPath: string;
}

/** Error types for merge operations. */
export class MergeError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'MergeError';
  }
}

export class InvalidGlobPatternError extends MergeError {
  constructor(pattern: string, source: Error) {
    super(`Invalid glob pattern '${pattern}': ${source.message}`);
    this.name = 'InvalidGlobPatternError';
  }
}

export class NoMatchesError extends MergeError {
  constructor(pattern: string) {
    super(`No files matched pattern '${pattern}'`);
    this.name = 'NoMatchesError';
  }
}

export class UnrecognizedFilenameError extends MergeError {
  constructor(filepath: string) {
    super(`Unrecognized transcript filename '${filepath}'`);
    this.name = 'UnrecognizedFilenameError';
  }
}

export class MixedDatesError extends MergeError {
  constructor() {
    super('Files correspond to multiple dates; supply --output to choose the destination');
    this.name = 'MixedDatesError';
  }
}

export class UndeterminedDateError extends MergeError {
  constructor() {
    super('Unable to determine an output filename; rerun with --output <file>');
    this.name = 'UndeterminedDateError';
  }
}

export class FileTooLargeError extends MergeError {
  constructor(filepath: string, size: number, max: number) {
    super(`File too large: ${filepath} (${size} bytes exceeds maximum of ${max} bytes)`);
    this.name = 'FileTooLargeError';
  }
}

export class IOError extends MergeError {
  constructor(message: string) {
    super(`I/O error: ${message}`);
    this.name = 'IOError';
  }
}

/** Sort key for chronological ordering of transcript files. */
interface FileSortKey {
  date: DateOnly | null;
  startHours: number;
  startMinutes: number;
  startSeconds: number;
}

/** Compare two sort keys for ordering. */
function compareSortKeys(a: FileSortKey, b: FileSortKey): number {
  // First compare by date if both have dates
  if (a.date && b.date) {
    if (a.date.year !== b.date.year) return a.date.year - b.date.year;
    if (a.date.month !== b.date.month) return a.date.month - b.date.month;
    if (a.date.day !== b.date.day) return a.date.day - b.date.day;
  } else if (a.date && !b.date) {
    return 1; // Files with dates come after files without
  } else if (!a.date && b.date) {
    return -1;
  }

  // Then compare by start time
  if (a.startHours !== b.startHours) return a.startHours - b.startHours;
  if (a.startMinutes !== b.startMinutes) return a.startMinutes - b.startMinutes;
  return a.startSeconds - b.startSeconds;
}

/** Check if filename matches flat format: YYYYMMDD_HHMMSS_HHMMSS */
function looksLikeFlatFormat(filename: string): boolean {
  const parts = filename.split('_');
  if (parts.length !== 3) return false;

  // Date part: 8 digits
  if (parts[0].length !== 8 || !/^\d{8}$/.test(parts[0])) return false;

  // Start time: 6 digits
  if (parts[1].length !== 6 || !/^\d{6}$/.test(parts[1])) return false;

  // End time: 6 digits
  if (parts[2].length !== 6 || !/^\d{6}$/.test(parts[2])) return false;

  return true;
}

/** Check if filename matches nested format: HHMMSS-HHMMSS */
function looksLikeNestedFormat(filename: string): boolean {
  const parts = filename.split('-');
  if (parts.length !== 2) return false;

  // Start time: 6 digits
  if (parts[0].length !== 6 || !/^\d{6}$/.test(parts[0])) return false;

  // End time: 6 digits
  if (parts[1].length !== 6 || !/^\d{6}$/.test(parts[1])) return false;

  return true;
}

/** Parse 6 digits as time components. */
function parseTimeDigits(value: string): { hours: number; minutes: number; seconds: number } | null {
  if (value.length !== 6 || !/^\d{6}$/.test(value)) return null;

  const hours = parseInt(value.slice(0, 2), 10);
  const minutes = parseInt(value.slice(2, 4), 10);
  const seconds = parseInt(value.slice(4, 6), 10);

  if (hours > 23 || minutes > 59 || seconds > 59) return null;

  return { hours, minutes, seconds };
}

/** Extract date from path directory structure (YYYY/MM/DD or YYYY-MM-DD). */
function extractDateFromPath(filepath: string): DateOnly | null {
  const dir = path.dirname(filepath);
  const dayDir = path.basename(dir);

  // Try YYYY-MM-DD format
  const dashedMatch = /^(\d{4})-(\d{2})-(\d{2})$/.exec(dayDir);
  if (dashedMatch) {
    return {
      year: parseInt(dashedMatch[1], 10),
      month: parseInt(dashedMatch[2], 10),
      day: parseInt(dashedMatch[3], 10),
    };
  }

  // Try nested YYYY/MM/DD format
  const parts = dir.split(path.sep);
  if (parts.length >= 3) {
    const dayPart = parts[parts.length - 1];
    const monthPart = parts[parts.length - 2];
    const yearPart = parts[parts.length - 3];

    if (
      /^\d{4}$/.test(yearPart) &&
      /^\d{2}$/.test(monthPart) &&
      /^\d{2}$/.test(dayPart)
    ) {
      const year = parseInt(yearPart, 10);
      const month = parseInt(monthPart, 10);
      const day = parseInt(dayPart, 10);

      // Validate date
      if (month >= 1 && month <= 12 && day >= 1 && day <= 31) {
        return { year, month, day };
      }
    }
  }

  return null;
}

/** Parse nested format (HHMMSS-HHMMSS). */
function parseNested(filepath: string, filename: string): FileSortKey {
  const parts = filename.split('-');
  const startTime = parseTimeDigits(parts[0]);

  if (!startTime) {
    throw new UnrecognizedFilenameError(filepath);
  }

  const date = extractDateFromPath(filepath);

  return {
    date,
    startHours: startTime.hours,
    startMinutes: startTime.minutes,
    startSeconds: startTime.seconds,
  };
}

/** Parse flat format (YYYYMMDD_HHMMSS_HHMMSS). */
function parseFlat(filepath: string, filename: string): FileSortKey {
  const parts = filename.split('_');
  const datePart = parts[0];
  const startPart = parts[1];

  const dateMatch = DATE_COMPACT_REGEX.exec(datePart);
  if (!dateMatch) {
    throw new UnrecognizedFilenameError(filepath);
  }

  const date: DateOnly = {
    year: parseInt(dateMatch[1], 10),
    month: parseInt(dateMatch[2], 10),
    day: parseInt(dateMatch[3], 10),
  };

  const startTime = parseTimeDigits(startPart);
  if (!startTime) {
    throw new UnrecognizedFilenameError(filepath);
  }

  return {
    date,
    startHours: startTime.hours,
    startMinutes: startTime.minutes,
    startSeconds: startTime.seconds,
  };
}

/** Extract sort key from file path. */
function getSortKey(filepath: string): FileSortKey {
  const basename = path.basename(filepath);
  const filename = basename.replace(/\.[^.]+$/, ''); // Remove extension

  // Try flat format first (more specific)
  if (looksLikeFlatFormat(filename)) {
    return parseFlat(filepath, filename);
  }

  // Try nested format
  if (looksLikeNestedFormat(filename)) {
    return parseNested(filepath, filename);
  }

  throw new UnrecognizedFilenameError(filepath);
}

/** Extract the day directory info for a nested file. */
function extractNestedDayDirectory(filepath: string): { dir: string; date: DateOnly } | null {
  const date = extractDateFromPath(filepath);
  if (!date) return null;
  return { dir: path.dirname(filepath), date };
}

/** Detect if all files share a common nested directory. */
function detectCommonNestedDirectory(paths: string[]): { dir: string; date: DateOnly } | null {
  let candidate: { dir: string; date: DateOnly } | null = null;

  for (const filepath of paths) {
    const info = extractNestedDayDirectory(filepath);
    if (!info) return null;

    if (candidate === null) {
      candidate = info;
    } else if (candidate.dir !== info.dir ||
               candidate.date.year !== info.date.year ||
               candidate.date.month !== info.date.month ||
               candidate.date.day !== info.date.day) {
      return null;
    }
  }

  return candidate;
}

/** Determine the output path for merged file. */
function determineOutputPath(
  ordered: string[],
  descriptors: Array<{ path: string; key: FileSortKey }>,
  request: MergeRequest
): string {
  // Use explicit output if provided
  if (request.output) {
    return request.output;
  }

  // Try to detect common nested directory
  const common = detectCommonNestedDirectory(ordered);
  if (common) {
    return path.join(common.dir, `${formatDateDashed(common.date)}.txt`);
  }

  // Check if all files have the same date
  let selectedDate: DateOnly | null = null;
  for (const { key } of descriptors) {
    if (key.date) {
      if (selectedDate === null) {
        selectedDate = key.date;
      } else if (
        selectedDate.year !== key.date.year ||
        selectedDate.month !== key.date.month ||
        selectedDate.day !== key.date.day
      ) {
        throw new MixedDatesError();
      }
    }
  }

  if (selectedDate) {
    const baseDir = ordered.length > 0 ? path.dirname(ordered[0]) : '.';
    return path.join(baseDir, `${formatDateDashed(selectedDate)}.txt`);
  }

  throw new UndeterminedDateError();
}

/** Write merged content to output file. */
function writeMergedFile(files: string[], outputPath: string): void {
  let merged = '';

  for (let i = 0; i < files.length; i++) {
    const content = fs.readFileSync(files[i], 'utf-8');
    merged += content;
    // Add newline between segments if needed
    if (i + 1 < files.length && !merged.endsWith('\n')) {
      merged += '\n';
    }
  }

  // Create parent directories
  const parentDir = path.dirname(outputPath);
  if (parentDir && parentDir !== '.') {
    fs.mkdirSync(parentDir, { recursive: true });
  }

  atomicWrite(outputPath, merged);
}

/** Delete source files after merge. */
function deleteSources(files: string[], outputPath: string): void {
  const outputCanonical = fs.existsSync(outputPath)
    ? fs.realpathSync(outputPath)
    : path.resolve(outputPath);

  for (const filepath of files) {
    const fileCanonical = fs.existsSync(filepath)
      ? fs.realpathSync(filepath)
      : path.resolve(filepath);

    // Never delete the output file
    if (fileCanonical === outputCanonical) {
      continue;
    }

    fs.unlinkSync(filepath);
  }
}

/** Execute the merge operation on transcript files. */
export async function executeMerge(request: MergeRequest): Promise<MergeOutcome> {
  const collected: string[] = [];

  // Expand glob patterns
  for (const pattern of request.patterns) {
    let matches: string[];
    try {
      matches = await glob(pattern, { nodir: true });
    } catch (error) {
      throw new InvalidGlobPatternError(
        pattern,
        error instanceof Error ? error : new Error(String(error))
      );
    }

    if (matches.length === 0) {
      throw new NoMatchesError(pattern);
    }

    collected.push(...matches);
  }

  // Check file sizes
  for (const filepath of collected) {
    let stats: fs.Stats;
    try {
      stats = fs.statSync(filepath);
    } catch (error) {
      throw new IOError(error instanceof Error ? error.message : String(error));
    }

    if (stats.size > MAX_FILE_SIZE) {
      throw new FileTooLargeError(filepath, stats.size, MAX_FILE_SIZE);
    }
  }

  // Extract sort keys and sort
  const descriptors: Array<{ path: string; key: FileSortKey }> = collected.map((filepath) => ({
    path: filepath,
    key: getSortKey(filepath),
  }));

  descriptors.sort((a, b) => compareSortKeys(a.key, b.key));

  // Deduplicate
  const ordered: string[] = [];
  for (const { path: filepath } of descriptors) {
    if (ordered.length === 0 || ordered[ordered.length - 1] !== filepath) {
      ordered.push(filepath);
    }
  }

  // Determine output path
  const outputPath = determineOutputPath(ordered, descriptors, request);

  // Canonicalize output path for reliable comparison
  const outputCanonical = fs.existsSync(outputPath)
    ? fs.realpathSync(outputPath)
    : path.resolve(outputPath);

  // Filter out the output path from sources
  const sourcesToMerge = ordered.filter((filepath) => {
    const fileCanonical = fs.existsSync(filepath)
      ? fs.realpathSync(filepath)
      : path.resolve(filepath);
    return fileCanonical !== outputCanonical;
  });

  // Write merged file
  try {
    writeMergedFile(sourcesToMerge, outputPath);
  } catch (error) {
    throw new IOError(error instanceof Error ? error.message : String(error));
  }

  // Delete sources if requested
  if (!request.noDelete) {
    try {
      deleteSources(sourcesToMerge, outputPath);
    } catch (error) {
      throw new IOError(error instanceof Error ? error.message : String(error));
    }
  }

  return {
    files: sourcesToMerge,
    outputPath,
  };
}

// Export helper functions for testing
export const _internal = {
  looksLikeFlatFormat,
  looksLikeNestedFormat,
  parseTimeDigits,
  getSortKey,
};
